src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceVariableService.cs

```diff
@@ -293,16 +299,21 @@ private CodeAction CreateAction(State state, bool allOccurrences, bool isConstan
             SemanticDocument currentDocument,
             SyntaxNode withinNodeInCurrent,
             bool allOccurrences,
+            bool includeLValues,
             CancellationToken cancellationToken)
         {
             var syntaxFacts = currentDocument.Project.LanguageServices.GetService<ISyntaxFactsService>();
+            var semanticFacts = currentDocument.Project.LanguageServices.GetService<ISemanticFactsService>();
             var originalSemanticModel = originalDocument.SemanticModel;
             var currentSemanticModel = currentDocument.SemanticModel;
 
             var result = new HashSet<TExpressionSyntax>();
             var matches = from nodeInCurrent in withinNodeInCurrent.DescendantNodesAndSelf().OfType<TExpressionSyntax>()
                           where NodeMatchesExpression(originalSemanticModel, currentSemanticModel, expressionInOriginal, nodeInCurrent, allOccurrences, cancellationToken)
-                          select nodeInCurrent;
+                            && (includeLValues                                   // if including lvalues too, everything passes, since rvalues always pass
+                                || (!includeLValues && !allOccurrences)          // if not including all occurences, including lvalues is handled earlier (ie rejected)
+                                || (!includeLValues && allOccurrences && semanticFacts.CanReplaceWithRValue(currentSemanticModel, nodeInCurrent, cancellationToken)))
+                          select nodeInCurrent;                                  // if not including lvalues but replacing all occurences, use CanReplaceWithRValue to determine if we can make the replacement (returns false for lvalues)
```

Comment:
this is too complex to undersstand at this point (esp. apparent with needing to have interstitial comments in a large expression).  Please just extract out a helper function, and use simple statements that are documented.  The code will be much easier to read, understand and maintain, and it will have the same perf characteristics.